Refactor resolver logic and add support for splitting types into separate files

hayes opened the pull request:
This was originally to resolve #142 but has since evolved into a more generic PR to allow for splitting up the generated code into multiple files.

## Overview

The goal of this PR was to enable splitting generated code for a single thrift namespace into multiple files.  For this to work, we needed a new way to resolve identifiers.  The current method of resolving identifiers was based an assumptions about all identifiers for the current namespace being available in the same file. 

#### Commits:
I broke out the changes into a set of commits with related changes because the PR is so large, it's hard to parse as a whole.

* Code changes:
  https://github.com/creditkarma/thrift-typescript/pull/143/commits/f1833b1b8aa220a85c235ebcbf8c4f4ed56e6bc7
* Resolver tests:
   https://github.com/creditkarma/thrift-typescript/pull/143/commits/448310fa851dda6922164efe33787d8cf1bc3ab4
* Validator tests:
  https://github.com/creditkarma/thrift-typescript/pull/143/commits/8ccb7a21fd30511507aa43ac241a1797afa2e422
* Codegen tests:
  https://github.com/creditkarma/thrift-typescript/pull/143/commits/beafad972fbe3b68e16b00f9bbf6aa26a728dbf9
### Resolver redesign


The new resolver is based on a set of classes that are designed to model the relationships between files and namespaces.  At the root of the resolver there is a new ResolverSchema class that has references to ResolverFiles and ResolverNames for all of the thrift files currently being parsed.  ResolverNamespaces map to the `js` (or fallback) namespaces, and contain references to each file that belongs to that namespace.

ResolverFiles are a representation of the underlying thrift files, and have references to the namespace they belong to, and to the other files that they have include statements for, and all the identifiers that are defined in each file.

By modeling the relationships between files in the thrift schema, we can invert the control flow for how identifiers are resolved, and greatly simplify the problem.  Previously, we needed to traverse the entire thrift AST to find all the identifier, and build complicated mappings of identifiers, which were then passed throughout the generator logic. This new representation can be built up easily by traversing just the top level nodes in the AST. 

The generator code has been updated to pass around the ResolverFile that corresponds to the statements that are being rendered, rather than the old identifier mappings.  When the generator code wants to resolve an identifier, it calls a method on the ResolverFile (`resolveIdentifier`) with the name of the identifier. 

This method can then resolve that identifier by looking at the types defined in the file, the includes in the file, and any global configuration for the schema.  This makes it easy to make changes to how an identifier is resolved depending on the context.

### File splitting

This was implemented by adding a new NamespaceGenerator class that is used to generate all the code for a given namespace.  It initially renders typescript statements for each type into a map, and then resolve those statements into either a single file, or a set of files (one per type).  When rendering into multiple files, the generator renders renders a set of import and export statements in the index file, so that all types are still exported the same way for either method. 

When rendering services using the multi-file strategy, no typescript namespace is generated.

hayes commented:
This current implementation is a bit hacky, and was just to try it out.  I've got an idea about how to handle this in a much better way that will allow easily splitting out anything into new files.  I'll update this PR in a couple hours with a better implementation

kevin-greene-ck commented:
Awesome, thanks. When ready, please add tests that give coverage of a feature like this. Something like: https://github.com/creditkarma/thrift-typescript/blob/master/src/tests/unit/index.spec.ts#L35

hayes commented on src/main/bin/resolveOptions.ts:
```diff
@@ -89,6 +89,11 @@ export function resolveOptions(args: Array<string>): IMakeOptions {
 
+            case '--file-per-type':
+                options.filePerType = true
+                index += 1
```
is this the right way to handle a boolean flag?

kevin-greene-ck replied:
Yeah, I think that's good. But if we go this way I'm not even sure we need the flag. This can just become the default behavior. It really shouldn't be breaking change for consumers.

hayes commented:
@kevinbgreene I've update this PR with a much more robust solution, that doesn't feel as much as a hack.  I haven't tested it yet, I'll work on that after lunch, but I think the over all structure is probably ready for feedback. 

This now adds a flag to render each type to its own file.

hayes commented:
Looks like I've got some work to do around resolving identifiers.

kevin-greene-ck commented:
> Looks like I've got some work to do around resolving identifiers.

Yeah, this is the problem. The way we're handling identifiers is not the best. I've wanted to refactor, but given everything is built on this it becomes a big ordeal.

hayes commented:
I was hoping to find something that didn't involve having to rework all of how identifiers work, but it's a pretty hard problem.  There are weird edge-cases like when thrift includes files from the same namespace.

Trying to figure out if I have time to actually rebuild the identifier logic

hayes commented:
Looks like there is no easy way to do this.  I've started working on a new way to resolve identifiers.  It's going to mean touching almost every file, which sucks.

If this seams like something that is unlikely to get approved, just let me know.

hayes commented:
This is far from complete at this point, but should give an idea of the direction I was working towards.

kevin-greene-ck commented:
@hayes Thanks man, this is definitely something I'll work with you to get approved. As I said, something I've been wanting to do. If you don't have time to get it across the finish line I can probably give it some time next week.

hayes commented:
Okay, I think I have the codegen working now.  Still need to update the validator and initialization logic.

hayes commented:
@kevin-greene-ck I think this is all working now.  I still need to update tests, which will be time consuming, but I've update the PR description with an overview of the changes, and tested this against a fairly large thrift codebase, and it seems to be working great.

hayes commented on src/main/generator/namespace.ts:
```diff
@@ -0,0 +1,174 @@
+                ...this.renderer.renderIncludes(
+                    this.namespace,
+                    this.options,
+                    '__NAMESPACE__',
```
This might not be the best name to use here, but works well enough for now.  Eventually, I would prefer to import just the referenced types from the current namespace (there shouldn't be any collisions when doing this).  But this PR is already really complicated and I didn't want to add that logic in the same PR

hayes commented on src/main/resolver/schema.ts:
```diff
@@ -0,0 +1,51 @@
+        const file = this.files.get(fileKey)
+
+        if (!file) {
+            throw new Error(`Unalb eto resolve file: ${fileName}`)
```
typo

hayes commented on src/tests/integration/thrift/common/common.thrift:
```diff
@@ -1,7 +1,7 @@
 namespace js common
 
-include "shared.thrift"
+include "../shared.thrift"
```
Does thrift somehow support includes with just the file name from a parent directory?  Need to check why this behavior changed.

hayes replied:
figured this out. works as it did before now.

hayes commented:
@kevin-greene-ck All tests are are passing now, I had to update a couple of them to work with the new resolver logic.  I also added tests for the new multi-file mode, and updated the generated tests to test both apache and thrift-server output.

hayes commented:
Looks like I forgot about one set of tests I had commented out.  Almost there!

hayes commented:
I'll be on vacation for the next week or so, but I think this is pretty much ready to go now.

kevin-greene-ck commented:
@hayes Cool, I started reviewing over the weekend. I'll get this merged and we'll commit any changes fixes we notice. Thanks for the great work.

hayes commented:
Any update on this?

kevin-greene-ck commented:
@hayes Hey, still looking through some things, but I'm going to put out a prerelease of `3.2.x` if you need it for development.

kevin-greene-ck commented:
Released this in `v3.2.0-0` will update this thread when it makes a standard release.

hayes commented:
Thanks!
